Dynamic_Programming/longest-increasing-subsequence.md

```diff
@@ -0,0 +1,253 @@
+- Arrays に fill() というメソッドがあったので使ってみる
+
+```java
+/**
+ * 時間計算量: O(n^2):
+ *     - O(n): キャッシュ用配列の作成
+ *     - O(n^2): 各要素とそれ以降のキャッシュすべてを利用した sequence 算出処理
+ *     - O(n): キャッシュから longestIncreasingSubsequence を探す処理
+ * 空間計算量: O(n)
+ *     - O(n): 引数 nums と同じサイズのキャッシュ用配列
+ */
+class Solution {
+    public int lengthOfLIS(int[] nums) {
+        if (nums == null || nums.length == 0) {
+            return 0;
+        }
+
+        int[] sequenceLengthCache = new int[nums.length];
+        Arrays.fill(sequenceLengthCache, 1);
+        for (int currentIndex = sequenceLengthCache.length - 1; currentIndex >= 0; currentIndex--) {
+            for (int subsequentIndex = currentIndex + 1; subsequentIndex < sequenceLengthCache.length; subsequentIndex++) {
+                if (nums[currentIndex] < nums[subsequentIndex]) {
+                    sequenceLengthCache[currentIndex] = Math.max(
+                        sequenceLengthCache[currentIndex],
+                        1 + sequenceLengthCache[subsequentIndex]
+                    );
+                }
+            }
+        }
+        int longestIncreasingSubsequence = 1;
+        for (int i = 0; i < sequenceLengthCache.length; i++) {
+            longestIncreasingSubsequence = Math.max(
+                longestIncreasingSubsequence,
+                sequenceLengthCache[i]
+            );
+        }
+        return longestIncreasingSubsequence;
+    }
+}
+```
+
+### 対応するインデックスの要素で終わる最長増加部分列の長さを配列に入れる
+
+他の方も悩まれてましたが、配列と関数の命名が難しいですね... これも直感的かどうかあんまり自信ないので、ご意見頂けますと嬉しいです。
+
+```java
+/**
+ * 時間計算量: O(n^2):
+ *     - O(n): キャッシュ用配列の作成
+ *     - O(n^2): 各要素とそれ以降の特定要素までの最長増加部分列算出処理
+ *     - O(n): キャッシュから longestIncreasingSubsequence を探す処理
+ * 空間計算量: O(n)
+ *     - O(n): 引数 nums と同じサイズのキャッシュ用配列
+ */
+class Solution {
+    public int lengthOfLIS(int[] nums) {
+        if (nums == null || nums.length == 0) {
+            return 0;
+        }
+
+        int[] maxLengthAsTail = new int[nums.length];
+        Arrays.fill(maxLengthAsTail, 1);
+        for (int i = 1; i < nums.length; i++) {
+            maxLengthAsTail[i] = findMaxLengthWithNewTail(nums, maxLengthAsTail, i);
+        }
+
+        int longestIncreasingSubsequence = 1;
+        for (int i = 0; i < maxLengthAsTail.length; i++) {
+            longestIncreasingSubsequence = Math.max(longestIncreasingSubsequence, maxLengthAsTail[i]);
+        }
+        return longestIncreasingSubsequence;
+    }
+
+    private int findMaxLengthWithNewTail(int[] nums, int[] maxLengthAsTail, int tailIndex) {
+        int maxLength = 1;
+        for (int i = 0; i < tailIndex; i++) {
+            if (nums[i] < nums[tailIndex]) {
+                maxLength = Math.max(maxLength, maxLengthAsTail[i] + 1);
+            }
+        }
+        return maxLength;
+    }
+}
+```
+
+### 二分探索を用いた O(n log n) の解法
+
+Java の標準ライブラリの二分探索メソッド、返り値がどうなるか理解するのがしんどかった... 要素が見つからなかったことを示すために負の値にするのはいいんだけど、なんで挿入位置を 1-indexed の値で返すようにしたんだろう🤔
```

Comment:
> index of the search key, if it is contained in the array; otherwise, (-(insertion point) - 1). The insertion point is defined as the point at which the key would be inserted into the array: the index of the first element greater than the key, or a.length if all elements in the array are less than the specified key. Note that this guarantees that the return value will be >= 0 if and only if the key is found.

`-(insertion point) - 1`（1-indexedとは？）ですね。要素が見つからず挿入位置が0の時と要素が見つかった場合を区別するためではないですか。

Reply:
https://stackoverflow.com/a/57911407/7503647
こういう回答はありました。たしかに0の挿入位置を表現しないといけないのはありますね